fusequery/query/src/optimizers/optimizer_projection_push_down.rs

```diff
@@ -0,0 +1,447 @@
+                    &input,
+                    &new_required_columns,
+                    true
+                )?)
+            }))
+        }
+        PlanNode::AggregatorPartial(AggregatorPartialPlan {
+            aggr_expr,
+            group_expr,
+            input
+        }) => {
+            // Partial aggregate:
+            // Remove any aggregate expression that is not needed
+            // and construct the new set of columns
+            exprvec_to_column_names(group_expr, &mut new_required_columns)?;
+
+            // Gather all columns needed
+            let mut new_aggr_expr = Vec::new();
+            aggr_expr.iter().try_for_each(|expr| {
+                let name = expr_to_name(&expr)?;
+
+                if required_columns.contains(&name) {
+                    new_aggr_expr.push(expr.clone());
+                    new_required_columns.insert(name.clone());
+                    expr_to_column_names(expr, &mut new_required_columns)
+                } else {
+                    Ok(())
+                }
+            })?;
+
+            Ok(PlanNode::AggregatorPartial(AggregatorPartialPlan {
+                aggr_expr: new_aggr_expr,
+                group_expr: group_expr.clone(),
+                input: Arc::new(optimize_plan(
+                    optimizer,
+                    &input,
+                    &new_required_columns,
+                    has_projection
+                )?)
+            }))
+        }
+
+        PlanNode::ReadSource(ReadDataSourcePlan {
+            db,
+            table,
+            schema,
+            partitions,
+            statistics,
+            description
+        }) => {
+            let projected_schema = get_projected_schema(schema, required_columns, has_projection)?;
+
+            Ok(PlanNode::ReadSource(ReadDataSourcePlan {
+                db: db.to_string(),
+                table: table.to_string(),
+                schema: projected_schema,
+                partitions: partitions.clone(),
+                statistics: statistics.clone(),
+                description: description.to_string()
+            }))
+        }
+        PlanNode::Empty(_) => Ok(plan.clone()),
+        _ => {
+            let input = plan.input();
+            let new_input = optimize_plan(optimizer, &input, &required_columns, has_projection)?;
+            let mut cloned_plan = plan.clone();
+            cloned_plan.set_input(&new_input)?;
+            Ok(cloned_plan)
+        }
+    }
+}
+
+impl IOptimizer for ProjectionPushDownOptimizer {
+    fn name(&self) -> &str {
+        "ProjectionPushDown"
+    }
+
+    fn optimize(&mut self, plan: &PlanNode) -> Result<PlanNode> {
+        let required_columns = plan
+            .schema()
+            .fields()
+            .iter()
+            .map(|f| f.name().clone())
+            .collect::<HashSet<String>>();
+        let new_node = optimize_plan(self, plan, &required_columns, false)?;
+        Ok(new_node)
+    }
+}
+
+#[cfg(test)]
```

Comment:
it's better move the test to optimizer_projection_push_down_test.rs file, it's recommend

Reply:
Fixed.